### ISSUE TITLE
`Nar hash mismatch between nix-store --dump and nix db` in 1.6.1

### ISSUE THREAD

**Issue by roberth** [cachix/cachix#592]
I'm getting this error from the attached dump.

```console
$ CACHIX_AUTH_TOKEN=$(read) cachix push hercules-ci /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path
Pushing 1 paths (284 are already present) using zstd to cache hercules-ci ⏳

✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)
✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)
✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)
✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)
✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)
✓ /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path (4.50 MiB)

/nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path: Nar hash mismatch between nix-store --dump and nix db. You can repair db metadata by running as root: $ nix-store --verify --repair --check-contents%                                                                                                                
```

However, the store did not have any corrupted paths, and Nix's recorded hash indeed matches the dump.

<details><summary>Shell session show that the dump and recorded hash match</summary>

```console
$ nix-store -q --hash /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path                                                                                                                                                                                                                                    
sha256:1isvm6xvwpcibnv583xg5nrgzmlsfy3rc95nj4ms7vxcyw92crwb
$ nix-store --dump /nix/store/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path >swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin                                                                                                                                                                                     
$ sha256sum <swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin                                                                                                                                                                                                                                                    
8b672612f7acefa32b91b6249687779ad6ffb22daf0f54b65d915dbebba95bc7  -
$ nix-hash --type sha256 --to-base32 8b672612f7acefa32b91b6249687779ad6ffb22daf0f54b65d915dbebba95bc7                                                                                                                                                                                                            
1isvm6xvwpcibnv583xg5nrgzmlsfy3rc95nj4ms7vxcyw92crwb

```

</details>

It seems that Cachix internally produces a different dump than Nix.

Here's the dump:
[swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin.gz](https://github.com/cachix/cachix/files/13280582/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin.gz)

**Comment by domenkozar** [comment]
We'll need to run https://github.com/sandydoo/hnix-store-dump and compare the outputs, might be a bug nar parser.

**Comment by domenkozar** [comment]
```
wget https://github.com/cachix/cachix/files/13280582/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin.gz
gunzip swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin.gz
nix-store --restore system-path < ~/Downloads/swyxxabx3l42fj8q0076qlyr1p2rjg5r-system-path.bin
nix-store --add system-path
nix-store -q --hash /nix/store/6vcq4y6ra892sapvr380f6ahz0qk212n-system-path
sha256:1isvm6xvwpcibnv583xg5nrgzmlsfy3rc95nj4ms7vxcyw92crwb
```

**Comment by domenkozar** [comment]
it's https://github.com/haskell-nix/hnix-store/issues/213

**Comment by sandydoo** [comment]
PR: https://github.com/haskell-nix/hnix-store/pull/218

**Comment by sandydoo** [comment]
@roberth, here's a branch with the fix applied, in case it helps unblock you.

https://github.com/cachix/cachix/tree/fix/592

**Comment by domenkozar** [comment]
Fixed in 1.7
